gcc/cp/contracts.cc

```diff
@@ -1613,61 +1618,208 @@ build_contract_function_decls (tree d)
   set_contract_functions (d, pre, post);
 }
 
-static const char *
-get_contract_level_name (tree contract)
+/* Build a layout-compatible internal version of source location __impl type.  */
+static tree
+get_pseudo_source_location_impl_type ()
 {
-  if (CONTRACT_LITERAL_MODE_P (contract))
-    return "";
-  if (tree mode = CONTRACT_MODE (contract))
-    if (tree level = TREE_VALUE (mode))
-      return IDENTIFIER_POINTER (level);
-  return "default";
+  if (!pseudo_source_location_impl_type)
+    {
+      // first build the __impl layout equivalent type
+      tree fields = NULL_TREE;
+      /* Must match <source_location>:
+	 struct __impl
+	  {
+	      const char* _M_file_name;
+	      const char* _M_function_name;
+	      unsigned _M_line;
+	      unsigned _M_column;
+	  }; */
+      const tree types[] = { const_string_type_node,
+			    const_string_type_node,
+			    uint_least32_type_node,
+			    uint_least32_type_node };
+
+      for (tree type : types)
+      {
+	/* finish_builtin_struct wants fieldss chained in reverse.  */
+	tree next = build_decl (BUILTINS_LOCATION, FIELD_DECL,
+				NULL_TREE, type);
+	DECL_CHAIN (next) = fields;
+	fields = next;
+      }
+
+      iloc_sentinel ils (input_location);
+      input_location = BUILTINS_LOCATION;
+      pseudo_source_location_impl_type = make_class_type (RECORD_TYPE);
+      finish_builtin_struct (pseudo_source_location_impl_type,
+			     "__pseudo_source_location_impl_type",
+			     fields, NULL_TREE);
+      CLASSTYPE_AS_BASE (pseudo_source_location_impl_type)
+	= pseudo_source_location_impl_type;
+      DECL_CONTEXT (TYPE_NAME (pseudo_source_location_impl_type))
```

Comment:
if this is private to the contracts impl, I wonder why we don't make it anonymous... (maybe there's a reason)?
likewise for the other cases below

Reply:
set TREE_PUBLIC to false for the new types